Feature: libcib, based: Support CIB transactions

nrwahl2 opened the pull request:
This pull request adds a new feature: atomic CIB transactions. It adds `init_transaction()` and `end_transaction()` methods to `cib_api_operations_t`. Transactions are supported for all three variants of `cib_t` clients: native, file, and remote. Native and remote share an implementation.

The basic idea is that a client can initiate a transaction, append requests to the transaction, and either commit or discard the transaction. Requests that are appended (via the `cib_transaction` value of `enum cib_call_options`) get queued until the transaction is committed. When a transaction is committed, all requests are processed in FIFO order. If any request fails, the transaction is aborted, and the CIB is left in its initial state. The CIB is modified only if all requests in the transaction succeed.

Supported requests are those that meet the following conditions:
* can be processed synchronously (with any changes applied to a working CIB copy)
* are not queries
* do not involve other nodes
* do not affect the state of pacemaker-based itself

In other words, the same types of requests are supported both in CIB file requests and in transactions.

(The following paragraph applies only to native and remote clients.) Replies are sent when a request is added to a transaction, **not** when the request is processed. This means callbacks are run at the client side after `pacemaker-based` tries to add the request to the transaction. Notifications are also not sent for individual requests. However, replies and notifications work normally for "commit-transaction" requests, and replies work normally for "init-transaction" and "discard-transaction" requests (which don't modify the CIB).

nrwahl2 commented:
@kgaillot Feel free to take a look if you want. Not necessary though. Changes are coming, particularly in `based_commit_transaction()`.

kgaillot commented on daemons/based/based_transaction.c:
```diff
@@ -0,0 +1,318 @@
+
+    // If successful, keep the working copy as the active CIB
+    if (rc == pcmk_rc_ok) {
+        crm_trace("CIB transaction complete");
```
Worth throwing in the client ID

kgaillot commented on daemons/based/based_transaction.c:
```diff
@@ -0,0 +1,318 @@
+     *   a callback and/or notification if applicable. Also cib_inhibit_notify
+     *   appears to affect only diff notifications, not replace notifications.
+     *   Perhaps do a dry-run of each request with cib_discard_reply and
+     *   cib_inhibit_notify.
```
I still think a single notification at the end is more appropriate. The key feature of transactions is atomicity, so it makes sense to extend that to the notification. It also avoids the problem of failed transactions. The only other solution I can see is queuing the notifications instead of sending them, and sending them all if the transaction completes successfully.

nrwahl2 replied:
Single at the end is probably for the best for notifications. Honestly, notifications were an afterthought that came to me right before I submitted this.

My original reasoning for "respond to each request individually" was for callbacks. I was thinking it might be useful to have a callback for a particular request -- especially for a request that ends up failing.

The only *specific, current* example I had in mind was using a callback for adding a request to the controller's expected replacements table, so that the CIB replacement wouldn't trigger an election. However, if we're able to avoid notifications for individual requests in a transaction, then we can just add the commit request to the expected replacements table.

I'll think about how to suppress notifications (and probably callbacks, unless I think of a way they're useful) for individual requests. The new call option idea below was quarter-baked in a hurry. We'd only need this internally, so maybe a function argument, a new member in some struct, or an XML field. I'll take a look tonight.

nrwahl2 replied:
I'm hoping to use `result_cib` to replace the CIB. That feels cleaner than creating a new `PCMK__CIB_REQUEST_REPLACE` request. The only downside is that `based_commit_transaction()` doesn't **completely** commit the transaction. It commits all the requests to a working CIB, but the commit isn't really finished until `cib_process_command()` checks the return code and activates `result_cib`.

Can't properly test until I finish resolving the other known issues/to-dos.

kgaillot reviewed (commented):
Love the general design, very clean

nrwahl2 commented on daemons/based/based_transaction.h:
```diff
@@ -0,0 +1,26 @@
+
+#include <crm_internal.h>
+
+#include <glib.h>
```
glib.h is not needed here anymore

nrwahl2 commented on daemons/based/pacemaker-based.h:
```diff
@@ -123,6 +125,14 @@ int cib_process_upgrade_server(const char *op, int options, const char *section,
                                xmlNode *req, xmlNode *input,
                                xmlNode *existing_cib, xmlNode **result_cib,
                                xmlNode **answer);
+int cib_process_init_transaction(const char *op, int options,
```
These could reasonably go in `based_transaction.h` but I kept them with the other `cib_process_*` functions.

nrwahl2 commented on daemons/based/based_transaction.c:
```diff
@@ -0,0 +1,318 @@
+        if (rc != pcmk_rc_ok) {
+            crm_err("Aborting CIB transaction due to failed '%s' request: %s",
+                    op, pcmk_rc_str(rc));
+            crm_log_xml_err(data->request, "Failed request");
```
Good point

nrwahl2 commented on daemons/based/based_transaction.c:
```diff
@@ -0,0 +1,318 @@
+     * can be applied. We can't achieve this guarantee for sync requests.
+     */
+    op = crm_element_value(request, F_CIB_OPERATION);
+    if (pcmk__str_any_of(op,
```
Also need to make sure `F_CIB_HOST` is not set to some other node.

It turns out that this request validation introduces a headache. If a developer submits an unsupported request while building a transaction, they won't know it was rejected unless they check the return code. Since the transaction was discarded, additional requests will be processed immediately instead of being queued. The first sign of trouble may be when they try to commit the transaction. By that time, they've accidentally applied changes.

I think it's better to add a new `cib_call_options` value called `cib_extend_transaction` or `cib_transaction` or something like that. This also gives the developer greater control. The developer will explicitly declare their intent to add a new request to a transaction, and they should check the return code to ensure that it was added successfully. If the `cib_extend_transaction` call option is **not** set, then the new request is processed immediately, and no attempt is made to add it to the client's transaction.

It's questionable whether an extension error should cause the transaction to be discarded immediately. An extension error should occur only in the event of a developer error (unsupported request) or a memory error. It could be convenient for all future calls with `cib_extend_transaction` to fail since no transaction exists, and for the commit to fail accordingly. That way, it's impossible for a transaction to be applied with a missing request due to a developer error. However, it may be better to let the developer decide how to respond to an extension failure (either by discarding the transaction or in some other way).

nrwahl2 replied:
On second thought, making non-extending calls while a transaction is open is probably a bad practice, leading to poor readability:
```
init_transaction(cib)
// return code checks elided
call1(cib, cib_extend_transaction)
call2(cib)
call3(cib, cib_extend_transaction)
commit_transaction(cib, cib_none)
```

nrwahl2 replied:
Darn... extending the transaction won't return a value to the client unless they call it with `cib_sync_call`. Will have to add everything to the transaction and do the validation at commit time.

kgaillot replied:
> I think it's better to add a new cib_call_options value called cib_extend_transaction or cib_transaction or something like that.

Definitely.

We could potentially drop the init transaction call, and just use the first call with cib_transaction to initialize a new transaction. Not sure whether that's more convenient or dangerous.

> It's questionable whether an extension error should cause the transaction to be discarded immediately. 

I could see it either way. On the one hand, the developer might want to commit any successful changes to that point, or try something different. On the other hand, atomicity would suggest any failure is a transaction failure.

> On second thought, devs should be checking the return code of each call anyway.

Definitely.

> And making non-extending calls while a transaction is open is probably a bad practice, leading to poor readability:

I don't think so. It's not necessarily in the same place; there could be a lot of async stuff going on, with a transaction being built gradually while normal calls proceed. Unlikely, but seems reasonable to support it.

> Darn... extending the transaction won't return a value to the client unless they call it with cib_sync_call. Will have to add everything to the transaction and do the validation at commit time.

pcmk_any_flags_set(flag, cib_sync_call|cib_transition)?

nrwahl2 replied:
> We could potentially drop the init transaction call, and just use the first call with cib_transaction to initialize a new transaction. Not sure whether that's more convenient or dangerous.

Mostly I like the symmetry (with commit) and clarity that `init_transaction()` gives. It also provides a bit of safety in case a developer forgets to commit a transaction and they intend to start a new one. If they do so via an "extends" call option, then it'll just be added to their existing transaction. If they do so via the init function, it will return `pcmk_rc_already`. (An unpublished early approach did use call options for everything BTW.)

> I could see it either way. On the one hand, the developer might want to commit any successful changes to that point, or try something different. On the other hand, atomicity would suggest any failure is a transaction failure.

The current draft does the validation at commit time. This turns out to be simpler: we treat an unsupported request exactly like a failed request. This comes at the expense of processing supported requests when an unsupported one is coming later, but that's not the common case.

> Unlikely, but seems reasonable to support it.

Yeah, I agree. In general I don't want to impose arbitrary restrictions. Another reason I chose to use functions rather than flags for init/commit is that the functions naturally allow trivial commits containing zero or one request.

We have enough non-arbitrary restrictions (for example, no calls that involve other nodes directly or indirectly).

> pcmk_any_flags_set(flag, cib_sync_call|cib_transition)?

Ah-ha, I'm not the only one who keeps typing "transition" out of habit.

Handling unsupported request failures at commit time renders this moot. This might work but would need to be done at the client side and feels a tad messier.

nrwahl2 replied:
> Handling unsupported request failures at commit time renders this moot.

Okay, that's not true. I still have this for `ENOMEM`:
```
            /* This should happen only on memory error. We have no way to notify
             * the client that extension failed (unless cib_sync_call was
             * given). Discard so that an incompletely built transaction is not
             * committed.
             */
            crm_err("Could not extend transaction for client %s, discarding: "
                    "%s",
                    pcmk__client_name(cib_client), pcmk_rc_str(rc));
            based_discard_transaction(cib_client);
```

Anyway, IIRC another reason was that sending a response to the client happens in `cib_process_request()` or `cib_process_command()`. We don't process a request that has `cib_extends_transaction` set. So even sending the reply from server side would get more complex.

nrwahl2 replied:
I decided it's best to at least send the reply at "extend transaction" time. An upcoming push attempts to do this. Checking the return code via callbacks and reacting it is super clunky, but it's best to have that option. The client can always make synchronous calls.

nrwahl2 commented:
I pushed a new iteration that addresses some wrinkles I found. I still haven't tested beyond `cts-regression`, because I've needed to fix the known issues first.

It's worth taking another look. We might be close, pending testing, inevitable bug fixes (hopefully nothing requiring major redesign), and the CIB-file implementation.

I'd especially appreciate a look at the `cib_inhibit_notify` and `cib_scope_local` commits. Those options are used for more than just transactions, so I want to be sure I get those right.

nrwahl2 commented on daemons/based/based_transaction.c:
```diff
@@ -0,0 +1,343 @@
+                  client_name, client_id);
+    }
+
+    /* Restore saved CIB. The caller activates *result_cib on success. This
```
fingers crossed :)

nrwahl2 commented:
Testing has begun (without file implementation yet).

Partially formed thought before bed: there *may* be a place for callbacks. The following CIB API functions grab output data (from `F_CIB_CALLDATA`):
* `noop()` (possible deprecation)
* `ping()` (possible deprecation)
* `query()`
* `query_from()`
* `erase()` (possible deprecation)

(`noop()`, `ping()`, and `erase()` are unused internally, but `noop()` is currently used by sbd; see [T656](https://projects.clusterlabs.org/T656). Not sure exactly what it's doing for sbd and whether we can replace it. `ping()` and `erase()` could be useful, we just aren't using them.)

We'd need a reply to the individual request in order to get that output data. We could simply make these requests unsupported within a transaction and avoid the issue that way. There may be limited use cases. For example, even though the transaction is atomic, the client that owns the transaction might want to query an intermediate state for some reason; it's a bit odd but not totally unreasonable.

`erase()` (or `PCMK__CIB_REQUEST_ERASE`) could be useful in a transaction, so I'd rather not drop support for erase requests in transactions. `erase()` can probably do its job fine without grabbing any output data, even though it's defined as doing so.

kgaillot commented on daemons/based/based_transaction.c:
```diff
@@ -0,0 +1,343 @@
+                          PCMK__CIB_REQUEST_QUERY,
+                          PCMK__CIB_REQUEST_REPLACE,
+                          PCMK__CIB_REQUEST_UPGRADE,
+                          NULL)) {
```
This should be a new boolean in cib_operation_t (actually the booleans should be flags, and there are other op checks that could be flags ...)

nrwahl2 replied:
Makes sense. I'll add a commit to convert to flags and then another to add a new flag for transaction support.

kgaillot commented on lib/cib/cib_client.c:
```diff
@@ -387,6 +387,30 @@ cib_client_erase(cib_t * cib, xmlNode ** output_data, int call_options)
+{
+    op_common(cib);
+    return cib_internal_op(cib, PCMK__CIB_REQUEST_DISCARD_TRANSACT, NULL, NULL,
+                           NULL, NULL, cib_none, NULL);
```
Separate commit/discard REQUESTs are fine, but just thinking out loud, an alternative would be a single end-transaction request with a boolean attribute for whether to commit.

nrwahl2 replied:
There's no input XML otherwise. So from a skim, we'd need to do one of:
* have separate commit/discard requests
* create some input XML just for the attribute
* add a new call option that only applies to the "end transaction" request

On the processing side, it seems reasonable to have separate functions for `based_discard_transaction()` and `based_commit_transaction()`, for clarity and to avoid an extra level of nesting.

kgaillot reviewed (commented):
Looking good. Just did a quick review at this point, need to look more closely at a couple of things later

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -114,7 +114,12 @@ void
+
+    if (xml == NULL) {
+        return;
+    }
```
True. In that case we need to add NULL checks to callers too. `create_xml_node()` is what got me down this tiny rabbit hole.

kgaillot replied:
How could this be NULL in create_xml_node()?

nrwahl2 replied:
If `xmlNewDocRawNode()` returns NULL. libxml2 doesn't assert on memory error like GLib does.

kgaillot replied:
We should definitely return NULL from create_xml_node() at that point rather than continue

nrwahl2 replied:
Agreed, that's what I ended up doing (not pushed yet).

nrwahl2 commented on daemons/based/based_callbacks.c:
```diff
@@ -1300,7 +1300,9 @@ cib_process_command(xmlNode * request, xmlNode ** reply, xmlNode ** cib_diff, gb
     // Calculate the hash value of the section before the change
-    if (pcmk__str_eq(PCMK__CIB_REQUEST_REPLACE, op, pcmk__str_none)) {
+    if (!pcmk_any_flags_set(call_options, cib_dryrun|cib_inhibit_notify)
+        && pcmk__str_eq(op, PCMK__CIB_REQUEST_REPLACE, pcmk__str_none)) {
```
I believe we already do for the most part. We start it unconditionally the first time we ever call `cib_process_command()`. In future calls, restarting the timer is guarded by:
```
    if (rc == pcmk_ok && !pcmk_is_set(call_options, cib_dryrun)) {
```

nrwahl2 commented on daemons/based/based_callbacks.c:
```diff
@@ -528,8 +528,11 @@ parse_local_options_v2(const pcmk__client_t *cib_client, int call_type,
     if (cib_op_modifies(call_type)) {
-        if (pcmk__str_any_of(op, PCMK__CIB_REQUEST_PRIMARY,
-                             PCMK__CIB_REQUEST_SECONDARY, NULL)) {
+        if (pcmk_is_set(call_options, cib_scope_local)
```
Working on getting the early commits broken out into a separate PR. I'm running CTSlab. It turns out this one causes a whole bunch of
```
May 16 12:21:49	BadNews: May 16 12:21:19 fastvm-rhel-9-0-42 pacemaker-based[73860]:  notice: Local CIB 0.27.244.e377a0bc5b6774df5df515e07f2953fa differs from fastvm-rhel-9-0-44: 0.27.240.90650fca63694fdb350e07d30d7b38d9 0x562b2b354dc0
May 16 12:21:49	BadNews: May 16 12:21:19 fastvm-rhel-9-0-42 pacemaker-based[73860]:  notice: Local CIB 0.27.244.e377a0bc5b6774df5df515e07f2953fa differs from fastvm-rhel-9-0-45: 0.27.240.90650fca63694fdb350e07d30d7b38d9 0x562b2b974bf0
May 16 12:21:49	BadNews: May 16 12:21:19 fastvm-rhel-9-0-42 pacemaker-based[73860]:  notice: Local CIB 0.27.244.e377a0bc5b6774df5df515e07f2953fa differs from fastvm-rhel-9-0-46: 0.27.240.90650fca63694fdb350e07d30d7b38d9 0x562b2b98bd00
May 16 12:21:49	BadNews: May 16 12:21:19 fastvm-rhel-9-0-42 pacemaker-based[73860]:  notice: Local CIB 0.27.244.e377a0bc5b6774df5df515e07f2953fa differs from fastvm-rhel-9-0-43: 0.27.240.90650fca63694fdb350e07d30d7b38d9 0x562b2b95c020
```

(not just the occasional known-issue ones)

I'll have to look into it. Surely `cib_scope_local` should make a CIB change local-only... Maybe `cib_scope_local` is being misused somewhere, since we've never done this in the past.

nrwahl2 replied:
Okay, I think the best bet is to deprecate `cib_scope_local` rather than trying to "fix" or extend it. It was nothing but a hack for legacy mode, which forwarded messages with `host == NULL` unless `cib_scope_local` was set or the local node was primary.

A few modifying requests use it for various reasons, and those would have to be handled if we extended `cib_scope_local` for non-legacy mode. Deprecating it is easier.

nrwahl2 replied:
Addressed in #3105, where I moved a lot of the preliminary stuff to de-clutter this PR

nrwahl2 commented:
Rebased on current main. No other changes in this push.

nrwahl2 commented:
Rebased on current main. No substantive changes yet.

kgaillot commented on include/crm/cib/cib_types.h:
```diff
@@ -80,6 +80,10 @@ enum cib_call_options {
     cib_dryrun          = (1 << 9),
+
+    //! Add request to the client's CIB transaction instead of processing
+    cib_transaction     = (1 << 10),
```
This is cib_extend_transaction in the rest of the commits

nrwahl2 replied:
Thanks yeah, I fixed it locally after I rebased and pushed

nrwahl2 commented:
Pushed to prove I am indeed working.

Updates:
* Functionize creation of reply XML (new commit) (`create_cib_reply()`)
* Move `do_local_notify()` (new commit)
* Validate request at transaction-extend time
* Don't discard transaction automatically after failed extend
* Send reply (via `do_local_notify()`) after trying to extend transaction
* Use `cib_op_attr_transaction` flag to validate that a request type is supported
* Log error message based on `rc` instead of `reason` in init, commit, discard processor functions (for future-proofing)
* Add missing `cib_process_discard_transaction()`
* Remove `process_transaction_request()`.
    * Its `cib_discard_reply` option goes into `parse_local_options_v2()` logic.
    * Its `cib_inhibit_notify` option is replaced by explicit checks for `cib_transaction` in `cib_process_request`.
    * (Note that a request with `cib_transaction` can reach `cib_process_request()` only via a `based_commit_transaction()`, unless I'm wrong. All requests have to pass through the initial entry point of `cib_common_callback_worker()` on some node, and there they get passed to `based_extend_transaction()` instead of being processed.)
* Improve doxygen and commit messages

Need to:
* do more testing after making all these changes
* write the file implementation
* make sure init/commit/discard messages can't get forwarded to a peer
* look for any weirdness with remote CIB clients (like with origin/target validation, sending/receiving replies, etc.)

kgaillot commented on daemons/based/based_callbacks.c:
```diff
@@ -174,6 +174,85 @@ create_cib_reply(const char *op, const char *call_id, const char *client_id,
+    free_xml(notify->notify_src);
+    free(notify->client_id);
+    free(notify);
+}
```
FYI local_notify_destroy_callback() is accidentally moved in this commit and moved back in a later commit

nrwahl2 commented:
small update to address review so far

nrwahl2 commented:
Minimal changes, mostly reordering (commits from #3113 come first, and `cib_op_attr_{replaces,writes_through}` come before transaction stuff).
* Incorporate commits from #3113 that were not already present here
* New commit: "Refactor: based: Improve consistency in parse_local_options_v2()"
* Transaction requests get the new `cib_op_attr_local` flag
* Committing a transaction now finishes by syncing the updated CIB to all nodes (in the "support CIB transactions" commit). `cib_process_commit_transaction()` and `based_commit_transaction()` get some small updates to allow this.

A transaction commit can't be forwarded to all nodes and processed in parallel like most modifying requests, since it depends on the local client table and operates on a local working CIB copy until it completes successfully. So applying the changes everywhere requires some extra care.

nrwahl2 commented:
Rebased on current main and pulled in the updated commits from #3113

nrwahl2 commented:
second push was accidental, wrong commits

nrwahl2 commented:
Not quite working... some type of admin_epoch nonsense that I need to debug, at the very least. Hacked `cibadmin` for testing:
```
May 30 02:57:06.494 laptop pacemaker-based     [1622606] (based_commit_transaction@based_transaction.c:278) 	debug: Committing transaction for client b5f36aea-32bc-4212-967d-11fb0093b3bb (cibadmin)
May 30 02:57:06.494 laptop pacemaker-based     [1622606] (cib_process_request@based_callbacks.c:1100) 	trace: Processing local cib_erase operation from cibadmin/3 intended for primary
May 30 02:57:06.494 laptop pacemaker-based     [1622606] (cib_perform_op@cib_utils.c:282) 	error: admin_epoch went backwards: 2 -> 1 (Opts: 0x4400)
May 30 02:57:06.494 laptop pacemaker-based     [1622606] (log_info@output_log.c:245) 	warning: Bad Op   <cib_command __name__="cib_command" t="cib" cib_op="cib_erase" cib_callid="3" cib_callopt="1024" cib_clientid="b5f36aea-32bc-4212-967d-11fb0093b3bb" cib_clientname="cibadmin" acl_target="root" cib_user="root" src="laptop"/>
May 30 02:57:06.494 laptop pacemaker-based     [1622606] (pcmk_log_xml_impl@logging.c:1142) 	warning: Bad Data No data to dump as XML
May 30 02:57:06.495 laptop pacemaker-based     [1622606] (log_info@output_log.c:245) 	info: Diff: --- 2.64.4 2
May 30 02:57:06.495 laptop pacemaker-based     [1622606] (log_info@output_log.c:245) 	info: Diff: +++ 1.1.0 (null)
May 30 02:57:06.495 laptop pacemaker-based     [1622606] (log_info@output_log.c:245) 	info: ++ /:  <cib crm_feature_set="3.17.4" validate-with="pacemaker-3.9" epoch="1" num_updates="0" admin_epoch="1"/>
May 30 02:57:06.495 laptop pacemaker-based     [1622606] (log_info@output_log.c:245) 	info: ++       <configuration>
May 30 02:57:06.495 laptop pacemaker-based     [1622606] (log_info@output_log.c:245) 	info: ++         <crm_config/>
May 30 02:57:06.495 laptop pacemaker-based     [1622606] (log_info@output_log.c:245) 	info: ++         <nodes/>
May 30 02:57:06.495 laptop pacemaker-based     [1622606] (log_info@output_log.c:245) 	info: ++         <resources/>
May 30 02:57:06.495 laptop pacemaker-based     [1622606] (log_info@output_log.c:245) 	info: ++         <constraints/>
May 30 02:57:06.495 laptop pacemaker-based     [1622606] (log_info@output_log.c:245) 	info: ++       </configuration>
May 30 02:57:06.495 laptop pacemaker-based     [1622606] (log_info@output_log.c:245) 	info: ++       <status/>
May 30 02:57:06.495 laptop pacemaker-based     [1622606] (log_info@output_log.c:245) 	info: ++     </cib>
May 30 02:57:06.495 laptop pacemaker-based     [1622606] (cib_process_command@based_callbacks.c:1588) 	trace: Not activating 0 0 0
```

nrwahl2 commented:
This finally seems to work. I've tested the basic transaction functionality and requirements that come to mind and don't see issues.
* [x] Transaction initiation fails if a transaction already exists
* [x]  Transaction extension fails if no transaction exists
* [x]  Transaction commit fails if no transaction exists
* [x]  Discard works
* [x]  A client's transaction gets freed on disconnect
* [x]  Entire transaction (multiple requests) is committed successfully if all requests succeed.
* [x]  A failed request causes failure with no side effects; transaction is discarded
* [x]  An unsupported request causes extend failure (EOPNOTSUPP) with no side effects; transaction is NOT discarded
* [x]  The CIB is replaced on success
* [x]  A replace notification is received on success (only one for the whole transaction)
* [x]  A diff notification is received (only one for the whole transaction, includes all changes)
* [x]  Callbacks are sent for individual requests at extend time, not at commit time

I still need to do the file implementation.

nrwahl2 commented on daemons/based/based_callbacks.c:
```diff
@@ -1448,16 +1491,26 @@ cib_process_command(xmlNode *request, const cib_operation_t *operation,
+         * call chain, and we need to accumulate changes in it.
+         *
+         * Status updates: copying large CIBs accounts for a huge percentage of
+         * our CIB usage, and this avoids some of it.
```
It's not clear to me why it was safe to use `cib_zero_copy` with the status section, especially before the `cib_zero_copy` improvements in a previous commit of this PR.

Also, following those improvements, it might be possible to use `cib_zero_copy` with all replace operations and maybe others (e.g., erase), where we don't use the copy that we make... I'm not gonna check that right now.

kgaillot replied:
The status section is never written to disk, so no activation is needed

nrwahl2 replied:
Optionally writing to disk is part of activation, and yes, `cib_zero_copy` causes `new_cib == saved_cib` which currently breaks `activateCibXml()`. But that's easy to change in `activateCibXml()`, and `cib_zero_copy` does more than that. 

With `cib_zero_copy`, in `cib_perform_op()`, instead of copying the whole CIB, we operate on the live CIB. The upshot is that if we make some change to the CIB and then return an error, that erroneous change stays in the CIB. We don't recover, because we don't have a good copy to fall back to.

So what we're saying with the existing use of `cib_zero_copy` is "we don't care if the status section is flat-out wrong in memory." I'm wary of that.

kgaillot replied:
> Optionally writing to disk is part of activation, and yes, `cib_zero_copy` causes `new_cib == saved_cib` which currently breaks `activateCibXml()`. But that's easy to change in `activateCibXml()`, and `cib_zero_copy` does more than that.
> 
> With `cib_zero_copy`, in `cib_perform_op()`, instead of copying the whole CIB, we operate on the live CIB. The upshot is that if we make some change to the CIB and then return an error, that erroneous change stays in the CIB. We don't recover, because we don't have a good copy to fall back to.
> 
> So what we're saying with the existing use of `cib_zero_copy` is "we don't care if the status section is flat-out wrong in memory." I'm wary of that.

Only the configuration section is written to disk. The status section is always in-memory only. (The scheduler saves the entire CIB including status to its inputs directory, but the CIB itself never writes out status.)

kgaillot replied:
I don't think status changes can fail, because there's no schema to validate

nrwahl2 replied:
> Only the configuration section is written to disk. The status section is always in-memory only. (The scheduler saves the entire CIB including status to its inputs directory, but the CIB itself never writes out status.)

Yeah, I didn't intend to suggest otherwise. I meant that optionally writing [something] to disk is part of activation, and that `cib_zero_copy` controls other things too.

My concern is what happens if the status section is wrong *in memory*. We don't know if all four of the assumptions I mentioned above hold. (We can feel pretty good about the ACL one, as you explained.) If something requests status info from based, it could grab inaccurate info from an earlier erroneous update with `cib_zero_copy`.

I'm wary of dropping this now, because based on the comment it helped a lot with performance. However, it looks like there are some tricky and rare bugs lurking -- and potential future bugs, if new `cib_op_t` functions aren't atomic.

nrwahl2 commented on daemons/based/based_callbacks.c:
```diff
@@ -1522,6 +1577,18 @@ cib_process_command(xmlNode *request, const cib_operation_t *operation,
+         */
+        if (pcmk__str_eq(op, PCMK__CIB_REQUEST_COMMIT_TRANSACT,
+                         pcmk__str_none)) {
+            sync_our_cib(request, TRUE);
```
Might be good to error-check the sync, though it's not clear how practical that is -- we've already activated the new CIB by this point. We could hack around it but it might be ugly.

kgaillot commented on daemons/based/based_callbacks.c:
```diff
@@ -274,6 +275,28 @@ cib_common_callback_worker(uint32_t id, uint32_t flags, xmlNode * op_request,
+
+        if (rc != pcmk_rc_ok) {
+            crm_err("Could not extend transaction for client %s: %s",
+                    pcmk__client_name(cib_client), pcmk_rc_str(rc));
```
I lean to warning for issues that only affect the client and are likely due to client input/behavior rather than the server failing at something

nrwahl2 replied:
Makes sense, will update

kgaillot commented on daemons/based/based_callbacks.c:
```diff
@@ -1069,6 +1107,11 @@ cib_process_request(xmlNode *request, gboolean privileged,
 
+    if (pcmk_is_set(call_options, cib_transaction)) {
+        CRM_CHECK(pcmk_is_set(operation->flags, cib_op_attr_transaction),
+                  return -EOPNOTSUPP);
```
Probably not a CHECK ... it's purely incorrect usage by client, right?

nrwahl2 replied:
No, this is a server issue. The server (`based_extend_transaction()`) should have rejected it at extend time. If we reach this point, it means we're in the process of committing a transaction. So we allowed an unsupported request through and we have a bug.

I'll add a comment.

kgaillot commented on daemons/based/based_callbacks.c:
```diff
@@ -1448,16 +1491,26 @@ cib_process_command(xmlNode *request, const cib_operation_t *operation,
     if (!pcmk_is_set(call_options, cib_dryrun)
-        && pcmk__str_eq(section, XML_CIB_TAG_STATUS, pcmk__str_casei)) {
-        // Copying large CIBs accounts for a huge percentage of our CIB usage
+        && !pcmk__str_eq(op, PCMK__CIB_REQUEST_COMMIT_TRANSACT, pcmk__str_none)
```
Do we need this if we're checking cib_transaction?

nrwahl2 replied:
This is mostly for future-proofing. `cib_dryrun` or `PCMK__CIB_REQUEST_COMMIT_TRANSACT` true guarantees that `cib_zero_copy` gets cleared. If those are both false, then either `cib_transaction` set or the status section as the target guarantees that `cib_zero_copy` gets set.

`cib_transaction` and `PCMK__CIB_REQUEST_COMMIT_TRANSACT` are mutually exclusive*. Currently there's no way the target for commit-transaction could be the status section either. So this condition is really saying "no matter what gets added here in the future, we always need to make a copy (i.e., clear `cib_zero_copy`) for commit-transaction." Otherwise, we'd be writing to the live CIB, and we couldn't abort and roll back.

*A commit-transaction request doesn't support being run within a transaction. I originally went with `cib_extend_transaction` for the call option name, as an attempt at clarity. `cib_in_transaction` might be a good middle ground if we choose to change it.

nrwahl2 replied:
On a related note, we should probably deprecate `cib_zero_copy` and keep the same logic internally -- maybe move it to `cib_perform_op()` if doing so doesn't break `cib_file`. Currently, if the caller sets `cib_zero_copy`, we completely ignore it already. It's either set or cleared in `cib_process_command()` by this if-else. It would be dangerous to allow a caller to make this decision.

kgaillot replied:
agreed

kgaillot commented on daemons/based/based_transaction.c:
```diff
@@ -0,0 +1,380 @@
+    client_id = pcmk__s(client->id, "unknown");
+
+    crm_trace("Discarding transaction for client %s (%s) if one exists",
+              client_name, client_id);
```
Don't need this message since both cases log another

nrwahl2 replied:
True. I added the conditional messages at the last minute because the "Discarding..." one looked weird at the end of a successful commit.

kgaillot commented on daemons/based/based_transaction.c:
```diff
@@ -0,0 +1,380 @@
+    }
+
+    client_name = pcmk__s(client->name, "unspecified");
+    client_id = pcmk__s(client->id, "unknown");
```
This is certainly fine, but FYI we have pcmk__client_name()

nrwahl2 replied:
Thanks, forgot that. Updated

kgaillot commented on daemons/based/based_transaction.c:
```diff
@@ -0,0 +1,380 @@
+
+    crm_trace("Finished processing transaction requests for client %s (%s); "
+              "discarding request queue",
+              client_name, client_id);
```
I prefer to combine messages when possible since logs can get mind-numbingly large e.g. `(rc == pcmk_rc_ok)? "Successfully" : "Unsuccessfully"`

nrwahl2 replied:
Good point. I added the "Finished processing" message later to explain why the "Discarding" message was happening.

Maybe "Transaction commit [succeeded|failed] for client `<name>` (`<id>`); discarding queue". We already have the "aborted" message with the RC string at error level if it failed.

kgaillot replied:
sounds good

kgaillot reviewed (commented):
Awesome

nrwahl2 commented:
Addressed review and pushed what I have so far for the `cib_file` implementation. It's almost finished-and-ready-for-testing.

I would love to merge `cib_server_ops` and `cib_file_ops` to drop that redundancy, but the differences in processor functions makes that a pain. The `based` functions aren't known to `libcib`. We could index into a function table (perhaps via an enum), but that's awfully similar to the `int call_type` stuff we got rid of in c13bf8b2. For now, I just duplicated things as needed.

nrwahl2 commented:
Alright, the file implementation is done. I performed the same tests as I performed for the `pacemaker-based` implementation, minus the ones that aren't applicable (e.g., callbacks and notifications).

This works AFAICT. Happy to discuss improvements.

nrwahl2 commented:
Also note that the only changes to the "Refactor: based: Support CIB transactions" commit were small ones to address review.

kgaillot commented on lib/cib/cib_file.c:
```diff
@@ -188,6 +196,29 @@ static gboolean cib_do_chown = FALSE;
+
+    if (*operation == NULL) {
+        crm_err("Operation %s is not valid", op);
+        return -EINVAL;
```
EOPNOTSUPP or EPROTO might be better

kgaillot replied:
My main thought is EINVAL maps to CRM_EX_SOFTWARE (i.e. bug) as exit code, while EOPNOTSUPP is CRM_EX_UNIMPLEMENT_FEATURE and EPROTO is CRM_EX_PROTOCOL. It might never be mapped to an exit status, but it still gives some logical direction. In this case, really any of the three would make logical sense, so I'm fine with any.

kgaillot commented on lib/cib/cib_file.c:
```diff
@@ -252,14 +283,9 @@ cib_file_perform_op_delegate(cib_t *cib, const char *op, const char *host,
-    }
-
-    if (operation == NULL) {
+    rc = get_operation(op, &operation);
```
the op==NULL check could move into get_operation()

nrwahl2 replied:
yeah, although that would change the return code (unless we special-case it below, inside the `if` block). Dunno if that even really matters.

kgaillot replied:
Or just return the desired final rc directly from get_operation()

nrwahl2 replied:
Good point. I was trying for a little while to figure out a good way to merge `cib_server_ops` and `cib_file_ops`, and I wanted the get-operation function to be identical in both implementations. Since that didn't work out, I can make them different.

nrwahl2 commented:
The "Drop internal use of cib_zero_copy" commit causes the seg faults on all non-Fedora, non-FreeBSD platforms.
* It keeps `pacemaker-based`'s "don't make a scratch copy if the target section is `status`" logic.
* It uses the same logic for both `pacemaker-based` and `cib_file` (via `cib_perform_op()`)
* `cib_file` has never set `cib_zero_copy` if the target is the status section.

So we get a seg fault on a replace operation (at least), when `section == XML_CIB_TAG_STATUS`.
 
As one example, the first test to fail is a `crm_ticket` test that deletes ticket state. `crm_ticket` replaces ticket state XML in the "status" section of the CIB.
```
cib_process_replace():
...
        xmlNode *obj_root = NULL;
        gboolean ok = TRUE;

        obj_root = pcmk_find_cib_element(*result_cib, section);
        ok = replace_xml_child(NULL, obj_root, input, FALSE);
        if (ok == FALSE) {
            crm_trace("No matching object to replace");
            result = -ENXIO;
        }

replace_xml_child():
...
        crm_log_xml_trace(child, "Delete match found...");
        if (delete_only || update == NULL) {
            free_xml(child);

        } else {
            xmlNode *tmp = copy_xml(update);
            xmlDoc *doc = tmp->doc;
            xmlNode *old = NULL;

            xml_accept_changes(tmp);
            old = xmlReplaceNode(child, tmp);

            if(xml_tracking_changes(tmp)) {
                /* Replaced sections may have included relevant ACLs */
                pcmk__apply_acl(tmp);
            }

            xml_calculate_changes(old, tmp);
            xmlDocSetRootElement(doc, old);
            free_xml(old);    <- seg fault on RHEL 9
```

kgaillot commented on lib/cib/Makefile.am:
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright 2004-2018 the Pacemaker project contributors
+# Copyright 2004-2023 the Pacemaker project contributors
```
I guess this is leftover from earlier version

nrwahl2 replied:
Yep, I started to make a `libcib_private.h` file and a `cib_file_transaction.c` file. Decided to leave that alone for now.

kgaillot reviewed (commented):
looks good

nrwahl2 commented:
Added four commits in the middle (before the `cib_zero_copy` commit) to address some XML issues (including a bug in `replace_xml_child()`. Forgot to address the review so I'll do that next

nrwahl2 commented:
One more push to address review and get rid of a testing commit that was accidentally added
